### PR TITLE
Pass runtime handler to image puller

### PR DESCRIPTION
**Pull request by wedsonaf**
This allows CRI servers to adjust their behaviour based on properties of the runtime class. For example, now that containerd allows each runtime to have its own snapshotter, it needs to know which runtime handler the kubelet intends to use before it can pull an image or check if it exists.

Background: we'd like to be able to run different runtime handlers (runc, kata, kata CC) with different snapshotters. Containerd added support for this via an annotation in the pod yaml to indicate the runtime handler.
Since different snapshotters may have different stores, this also affects image pulling.
If the pod pull policy is “IfNotPresent”, then the kubelet calls containerd’s image service to determine if an image is present or not (by calling the ImageStatus function) to decide whether to pull.
 
The issue:
Containerd doesn’t check the runtime handler annotation to determine if an image is present. When we fix this, then everything works well for annotated pods yamls.
But pod yamls that aren’t annotated break as follows: if we run an annotated pod first (which pulls the image using its snapshotter), then when we run a pod without annotations, the kubelet will think that the image is already present because it was fetched by another snapshotter, and kata fails to start.
If we fix this by changing containerd to consider that the runtime handler for a pod without annotations is the default handler, then we break tools like crictl. For example, when we run `crictl rmi -a` to remove all images, it fails on images that aren’t on the snapshotter for the default handler.
 
The solution:
Annotate all pods.
 
We have two issues with that:
1.	We’d like to able to continue to run runc and kata handler unchanged. This isn’t possible if we require users to annotate their pod yamls.
2.	Even for kata cc pod yamls, the annotations lack usability because they require the runtime handler (not the runtime class). By changing containerd alone we can’t fix this because runtime classes are a Kubernetes concept.
 
Proposal: have the kubelet translate the a pod’s runtime class to runtime handler and add the annotation to ImageSpec.
 
This has been discussed in containerd upstream, for example, https://github.com/containerd/containerd/issues/6657#issuecomment-1066478023 suggests adding this annotation to ImageSpec. There’s also a KEP (https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1301-windows-runtime-class) that talks about doing the same.

### CODE REVIEW

File: pkg/kubelet/images/image_manager.go
Comment (haircommander) [inline pkg/kubelet/images/image_manager.go:128]: can this be defined somewhere public so CRI implementations can import the variable? maybe in staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/constants.go

**Comment by haircommander** [comment]
/ok-to-test
/triage accepted
/priority important-longterm
/sig node

@wedsonaf you'll need to sign the CLA and sign-off the commit

File: pkg/kubelet/images/image_manager.go
Comment (wedsonaf) [reply pkg/kubelet/images/image_manager.go]: Thanks for the suggestion. 

Done in the latest version.

**Review by mikebrow — COMMENTED** [review]
See comments

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:179]: Sure, will do.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:185]: Ack.

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go:190]: > Maybe the best route is just to add the handler as an annotation in the pod spec, then the copy to the image services would be automatic by pod annotation copy.

Yes, I think this would simplify this patch. Let me take a stab at it.

File: staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/constants.go
Comment (mikebrow) [reply staging/src/k8s.io/cri-api/pkg/apis/runtime/v1/constants.go:59]: let's just leave it as a string as originally done..

File: pkg/kubelet/kuberuntime/kuberuntime_container.go
Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_container.go]: @mikebrow I just uploaded a version that I believe does what you suggest and is much simpler, just the addition of 7 lines overall, including test.

Please let me know what you think!

**Comment by wedsonaf** [comment]
> @wedsonaf: The following test **failed**, say `/retest` to rerun all failed tests or `/retest-required` to rerun all mandatory failed tests:
> 
> Test name	Commit	Details	Required	Rerun command
> pull-kubernetes-node-e2e-containerd	[a0b7bd8](https://github.com/kubernetes/kubernetes/commit/a0b7bd85f5d130e5073175916e3d143cb300733e)	[link](https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/118907/pull-kubernetes-node-e2e-containerd/1697331721587920896/)	true	`/test pull-kubernetes-node-e2e-containerd`
> [Full PR test history](https://prow.k8s.io/pr-history?org=kubernetes&repo=kubernetes&pr=118907). [Your PR dashboard](https://prow.k8s.io/pr?query=is%3Apr%20state%3Aopen%20author%3Awedsonaf). Please help us cut down on flakes by [linking to](https://git.k8s.io/community/contributors/devel/sig-testing/flaky-tests.md#filing-issues-for-flaky-tests) an [open issue](https://github.com/kubernetes/kubernetes/issues?q=is:issue+is:open) when you hit one in your PR.
> 
> Instructions for interacting with me using PR comments are available [here](https://git.k8s.io/community/contributors/guide/pull-requests.md). If you have questions or suggestions related to my behavior, please file an issue against the [kubernetes/test-infra](https://github.com/kubernetes/test-infra/issues/new?title=Prow%20issue:) repository. I understand the commands that are listed [here](https://go.k8s.io/bot-commands).

There are two failures. Neither seems related to this change (one has to do with being killed by the oom killer and the other starts off by failing to fetch a resource [`projects/k8s-infra-e2e-boskos-063/regions/us-west1/addresses/bootstrap-e2e-master-ip`]).

Looking at the job history, they both occur in other unrelated PR, so it looks like they are flaky tests. Here are the 3 most recent failures: https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120412/pull-kubernetes-node-e2e-containerd/1699064085817069568, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/120253/pull-kubernetes-node-e2e-containerd/1699063563743662080, https://prow.k8s.io/view/gs/kubernetes-jenkins/pr-logs/pull/119200/pull-kubernetes-node-e2e-containerd/1699051057583230976

@mikebrow -- would you be able to rerun the failing tests?

**Comment by wedsonaf** [comment]
> /test pull-kubernetes-node-e2e-containerd

Thanks @mikebrow -- the tests did pass this time around.

Any feedback on the PR? It's tiny now:

```
 pkg/kubelet/kuberuntime/kuberuntime_sandbox.go      | 6 ++++++
 pkg/kubelet/kuberuntime/kuberuntime_sandbox_test.go | 1 +
 2 files changed, 7 insertions(+)
 ```

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (mikebrow) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: let's only add the annotations map and annotation if runtimeHandler != ""

Which can be done above after the log on line 65..

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox_test.go
Comment (mikebrow) [inline pkg/kubelet/kuberuntime/kuberuntime_sandbox_test.go:170]: and since we are not going to force the annotations map into existence we need to do the assert for the annotation only if expected.. to avoid the nil ptr

**Review by mikebrow — COMMENTED** [review]
see comment :-)

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: I'm planning to use the presence of the annotation to determine how the cri server should behave.

For example, see this commit: https://github.com/microsoft/confidential-containers-containerd/commit/494207280f55c0ccd4c54ca7cfd24a8ae5d9fd09 (a variant of which I intend to include in a PR to containerd once we the kubelet starts adding the annotation).

The idea is that containerd would behave exactly as before without the annotation. If the annotation is present, even if runtimeHandler == "", it will return the image status based on the presence of a snapshot for the appropriate snapshotter.

This is something we discussed in that triage meeting some time back: we need the old behaviour when, for example, we want to do `crictl rmi` to remove an image from all snapshotters.

WDYT?

Comment (mikebrow) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: https://github.com/microsoft/confidential-containers-containerd/commit/494207280f55c0ccd4c54ca7cfd24a8ae5d9fd09#r127070837

Comment (mikebrow) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: if image.Annotations is == nil || runtimeHandler == "" then set runtimeHandler to c.config.ContainerdConfig.DefaultRuntimeName

Comment (aojea) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: will this override the annotation if already exist?

Comment (mikebrow) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: see link ^ ... here in kubelet you can leave annotations unset esp. when there are no annotations.. Over in your code just check for annotations being nil or if not nil then check for existence of the handler key/value pair

Comment (wedsonaf) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: In the beginning I had similar code to what you propose in the commit comment, however, that breaks `crictl rmi` and compatibility with kubelets that don't add runtime annotations.

The scenario is the following: you ran some container with a non-default runtime handler that has a different snapshotter, then you want to manually remove the image. If you run `crictl rmi X` or `crictl rmi -a`, it will fail because the snapshotter for the default runtime handler doesn't have a snapshot for the image (for it to work you'd have to add an annotation but the tool doesn't even support that yet for rmi like it does for pull).

Things also break in similar fashion if you pair this containerd with a kubelet that doesn't add runtime annotations: every time it tries to ask the status of an image for which the snapshotter of the default runtime doesn't have a snapshot, it will fail.

My solution to this issue, at a high level, is to preserve previous behaviour when the runtime handler annotation is absent, and only trigger new behaviour when it is present. This covers both scenarios above.

**Review by mikebrow — APPROVED** [review]
LGTM

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (mikebrow) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: agree with your logic.. LGTM

**Review by feiskyer — APPROVED** [review]
looks good

**Comment by wedsonaf** [comment]
@SergeyKanzhelev -- any chance you could look into this for approval?

We discussed it in a sig-node meeting some time back. After Mike's feedback, it's a tiny 7-line change including tests.

**Comment by miwithro** [comment]
@derekwaynecarr can we get an approval of the PR please?

**Comment by miwithro** [comment]
@dchen1107 can we get approval?  Thank you.

**Comment by kiashok** [comment]
> This allows CRI servers to adjust their behaviour based on properties of the runtime class. For example, now that containerd allows each runtime to have its own snapshotter, it needs to know which runtime handler the kubelet intends to use before it can pull an image or check if it exists.
> 
> Background: we'd like to be able to run different runtime handlers (runc, kata, kata CC) with different snapshotters. Containerd added support for this via an annotation in the pod yaml to indicate the runtime handler. Since different snapshotters may have different stores, this also affects image pulling. If the pod pull policy is “IfNotPresent”, then the kubelet calls containerd’s image service to determine if an image is present or not (by calling the ImageStatus function) to decide whether to pull.
> 
> The issue: Containerd doesn’t check the runtime handler annotation to determine if an image is present. When we fix this, then everything works well for annotated pods yamls. But pod yamls that aren’t annotated break as follows: if we run an annotated pod first (which pulls the image using its snapshotter), then when we run a pod without annotations, the kubelet will think that the image is already present because it was fetched by another snapshotter, and kata fails to start. If we fix this by changing containerd to consider that the runtime handler for a pod without annotations is the default handler, then we break tools like crictl. For example, when we run `crictl rmi -a` to remove all images, it fails on images that aren’t on the snapshotter for the default handler.
> 
> The solution: Annotate all pods.
> 
> We have two issues with that:
> 
> 1. We’d like to able to continue to run runc and kata handler unchanged. This isn’t possible if we require users to annotate their pod yamls.
> 2. Even for kata cc pod yamls, the annotations lack usability because they require the runtime handler (not the runtime class). By changing containerd alone we can’t fix this because runtime classes are a Kubernetes concept.
> 
> Proposal: have the kubelet translate the a pod’s runtime class to runtime handler and add the annotation to ImageSpec.
> 
> This has been discussed in containerd upstream, for example, [containerd/containerd#6657 (comment)](https://github.com/containerd/containerd/issues/6657#issuecomment-1066478023) suggests adding this annotation to ImageSpec. There’s also a KEP (https://github.com/kubernetes/enhancements/tree/master/keps/sig-windows/1301-windows-runtime-class) that talks about doing the same.

@wedsonaf I think we discussed this several months ago but wanted to reconfirm - are the changes here only a temporary solution till we have a dedicated field in ImageSpec to set the runtime handler? KEP that addresses this - https://github.com/kubernetes/enhancements/pull/4217

**Comment by wedsonaf** [comment]
> @wedsonaf I think we discussed this several months ago but wanted to reconfirm - are the changes here only a temporary solution till we have a dedicated field in ImageSpec to set the runtime handler? KEP that addresses this - [kubernetes/enhancements#4217](https://github.com/kubernetes/enhancements/pull/4217)

Yes, we discussed this several months ago.

This is a solution to address our current issues. If your proposal eventually makes it, we can use it as well instead.

**Comment by kiashok** [comment]
> > @wedsonaf I think we discussed this several months ago but wanted to reconfirm - are the changes here only a temporary solution till we have a dedicated field in ImageSpec to set the runtime handler? KEP that addresses this - [kubernetes/enhancements#4217](https://github.com/kubernetes/enhancements/pull/4217)
> 
> Yes, we discussed this several months ago.
> 
> This is a solution to address our current issues. If your proposal eventually makes it, we can use it as well instead.

sounds good, thanks!

cc @jsturtevant  @marosset

**Comment by jsturtevant** [comment]
> > This is a solution to address our current issues. If your proposal eventually makes it, we can use it as well instead.
> 
> sounds good, thanks!

@wedsonaf can you verify the approach outlined in the KEP would meet your requirements?  From what I can see here, it looks like it should.  

We've used the approach of annotations in the past as a stop gap until CRI changes and features land fully so this approach LGTM.  It will unblock you and it helps us be able to leverage the changes on the Windows contained side until the KEP makes it to a state where it is turned on by default.

**Comment by haircommander** [comment]
I feel like we should go with https://github.com/kubernetes/kubernetes/pull/121121 at this point, as the enhancement has been accepted and the approach agreed upon

**Comment by haircommander** [comment]
> I feel like we should go with https://github.com/kubernetes/kubernetes/pull/121121 at this point, as the enhancement has been accepted and the approach agreed upon

I've been reminded this is fixing a regression. We could go with this, revert in https://github.com/kubernetes/kubernetes/pull/121121 or instead just skip the master branch and start the backport process to fix in other branches while we move forward with the long-term solution

**Comment by derekwaynecarr** [comment]
@haircommander whether we go with this approach, or with the linked enhancement, is there not an issue with how the kubelet does image gc?  what happens when the image gc manager lists images?  does it see the same image twice?  in the image manager gc controller, would this confuse its image in use detection?

**Comment by mikebrow** [comment]
suggest adding into master and immediately deprecating... remove with eol for back level container runtimes

**Comment by haircommander** [comment]
> is there not an issue with how the kubelet does image gc?

@derekwaynecarr we talked about this in the image GC WG. The way we reasoned about it was: kubelet should see each image as a different one, as even though they share the manifest list, they'll have different digests and read as unique to the kubelet. Thus, kubelet will treat each image as separate, and issue a GC for each individual one. The RemoveImage command may end up being effectively a no-op in the CRI, as the entry in the manifest list may be deleted but the image not actually removed, but this happens today with images that share layers and are tagged as different names. So we're pretty sure GC should continue working as expected. This change just gives the runtime a chance to quickly find where the different images are stored. @kiashok did I get that summary right?

**Comment by kiashok** [comment]
> > is there not an issue with how the kubelet does image gc?
> 
> @derekwaynecarr we talked about this in the image GC WG. The way we reasoned about it was: kubelet should see each image as a different one, as even though they share the manifest list, they'll have different digests and read as unique to the kubelet. Thus, kubelet will treat each image as separate, and issue a GC for each individual one. The RemoveImage command may end up being effectively a no-op in the CRI, as the entry in the manifest list may be deleted but the image not actually removed, but this happens today with images that share layers and are tagged as different names. So we're pretty sure GC should continue working as expected. This change just gives the runtime a chance to quickly find where the different images are stored. @kiashok did I get that summary right?

kubelet GC currently relies only on the imageName/imageID while making a RemoveImage() call to the container runtime. With KEP 4216, we would need to change behavior of kubelet GC to also include the runtimeHandler used to pull the image with RemoveImage() call to container runtime. If runtimeHandler field is "", then container runtime like containerd would pick the default runtime handler for the image. Alpha changes for KEP 4216 under a feature gate is here: https://github.com/kubernetes/kubernetes/pull/121456

**Comment by skaegi** [comment]
Could we please add back the annotation as we need some overlap between kubernetes and containerd and our code. We use this in production to support two snapshotters on a node and although it hasn't currently caused a regression and outage in production it eventually will! We are onboard to switch to the "final" change when we move to 1.29 but would be good to have a few releases of overlap.

**Comment by fidencio** [comment]
Any chance to get some traction on this one?
/cc @haircommander  

While I fully understand the KEP is the correct way to have this fixed, this simple patch could be applied and easily backported to previous versions of k8s, and that would be a very good stop-gap till the KEP is fully implemented and we have massive adoption of the k8s version where the KEP will land.

File: pkg/kubelet/kuberuntime/kuberuntime_sandbox.go
Comment (bart0sh) [reply pkg/kubelet/kuberuntime/kuberuntime_sandbox.go:73]: @wedsonaf 
> will this override the annotation if already exist?

and should we avoid this?

**Comment by fidencio** [comment]
Folks, in order to address @bart0sh comments, I've created https://github.com/kubernetes/kubernetes/pull/124343, as @wedsonaf has been working on different things at this time.

I'd appreciate if we can move the discussions there, as I don't have the rights to force push to his branch.

**Comment by dims** [comment]
/close

please switch to https://github.com/kubernetes/kubernetes/pull/118907

**Comment by dims** [comment]
Apologies :) yes.

**Pull request closed without merging**
